### ISSUE TITLE
Istio CNI Agent: Tests around CNI plugin for sidecar-and-ambient-together scenarios

### ISSUE THREAD

**Issue by bleggett** [istio/istio#49228]
As per https://github.com/istio/istio/pull/49207 - we need more test coverage around using/enabling/upgrading sidecar and ambient at the same time in CNI machinery.

Probably missing some (expensive) integration tests, and more (cheap) unit tests.

**Affected product area (please put an X in all that apply)**

[x] Ambient
[ ] Docs
[ ] Dual Stack
[x] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[x] Test and Release
[ ] User Experience
[x] Developer Infrastructure

**Comment by aazam-gh** [comment]
Hello, can I take this up :)

**Comment by biswajit-9776** [comment]
Hello @bleggett I would like to add tests for this issue. Could you tell me how would the plugin behave with using/enabling/upgrading sidecar and ambient scenarios?

**Comment by bleggett** [comment]
@biswajit-9776 Behavior should be the same in a mixed environment for every kind of workload as it would be if "only sidecar" workloads were deployed, or "only ambient" workloads were deployed.

We test all that separately, but we need to run the basic scenarios we already have in a mixed (some sidecar, some ambient workloads) test env and make sure each kind of pod continues to work exactly like it would if the other pods/mode was not there.

**Comment by biswajit-9776** [comment]
The above PR added does not include many other test cases. I wanted to add another function for checking the CmdAdd for a pod with a sidecar injected without Ambient being enabled in namespace at all. My doubt was should the pod be added to mesh?

**Comment by bleggett** [comment]
@biswajit-9776 The unit test you added there is redundant with an existing unit test (thanks for jumping in tho!)

if you are interested in helping here, could you look at adding some more tests to the integration suite for Helm upgrades?

https://github.com/istio/istio/pull/49179 added the initial set - we could use some more tests there around testing upgrades when both ambient and sidecar are enabled.

**Comment by PavloGl** [comment]
@biswajit-9776 are you still working on this?

**Comment by biswajit-9776** [comment]
> @biswajit-9776 are you still working on this?

No, I am not working on this issue anymore.

**Comment by PavloGl** [comment]
@bleggett Can you assign this issue to me?

**Comment by kfaseela** [comment]
> @biswajit-9776 The unit test you added there is redundant with an existing unit test (thanks for jumping in tho!)
> 
> if you are interested in helping here, could you look at adding some more tests to the _integration_ suite for Helm upgrades?
> 
> #49179 added the initial set of ambient upgrade tests - we could use some more tests in that suite around testing upgrades when _both_ ambient and sidecar are enabled.

@bleggett : is someone already working on this? else i can pick this up

**Comment by bleggett** [comment]
@pglushko asked for it - if they are still working on it, maybe check with them.


More detail on the integration tests:

`istio-cni` is optional for sidecars, and required for ambient.

- I believe we have integration tests that test sidecar (not using `istio-cni`) and ambient together.
- I do _not_ believe we have integration tests that test sidecar (using `istio-cni`) and ambient together.

**Comment by kfaseela** [comment]
@pglushko let me know if you are already working on this. otherwise @adilmp1 would like to work on this.

**Comment by kfaseela** [comment]
Makes sense to have a separate suite enabling this in the installation and run TestTraffic alone?

**Comment by bleggett** [comment]
Yep, maybe `integration/ambient/repair_test.go` which installs with repair ON and makes sure nothing breaks.


We can have either

- repair default-OFF for the entire ambient test suite, and one test that validates with repair=ON 
- or the opposite, where repair is default-ON and we have one test that validates with repair=OFF

I don't care which, your choice, but both cases should be covered.

**Comment by kfaseela** [comment]
Should this be part of the same PR or a different one?

**Comment by bleggett** [comment]
> Should this be part of the same PR or a different one?

Ideally same, I guess? But I am ok with either.

**Comment by adilmp1** [comment]
#50673
After enabling CNI repair, I used TestTraffic as the test. Is that sufficient, @bleggett ?

**Comment by kss682** [comment]
I would like to work on this issue, if it is still relevant.

**Comment by kfaseela** [comment]
@kss682 : This work was already completed by @adilmp1
